### PR TITLE
fix(rust, python) respect time zone in date_range

### LINKED ISSUE #7418: Ambiguous time bug
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

In some functions in `polars` certain times with different timezone are considered `ambiguous` (e.g `2021-11-07T01:00:00CDT` and  `2021-11-07T01:00:00CST`) whereas they are not as the timezone is well specified. This happens in the following example but also in `groupby`.
I'm not sure why in some other function they are not considered `amiguous` though, e.g in `convert_time_zone` works fine.

``` python
>>> pl.date_range(datetime(2021,11,7), datetime(2021,11,8), "1h", time_zone="US/Central")
thread '<unnamed>' panicked at 'Ambiguous local time, ranging from 2021-11-07T01:00:00CDT to 2021-11-07T01:00:00CST', /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/chrono-0.4.23/src/offset/mod.rs:189:17
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/***/***/lib/python3.10/site-packages/polars/internals/functions.py", line 497, in date_range
    _py_date_range(start, stop, interval, closed, name, tu, time_zone)
pyo3_runtime.PanicException: Ambiguous local time, ranging from 2021-11-07T01:00:00CDT to 2021-11-07T01:00:00CST
>>> pl.date_range(datetime(2021,11,7), datetime(2021,11,8), "1h", time_zone="UTC").dt.convert_time_zone("US/Central")
shape: (25,)
Series: '' [datetime[μs, US/Central]]
[
	2021-11-06 19:00:00 CDT
	2021-11-06 20:00:00 CDT
	2021-11-06 21:00:00 CDT
	2021-11-06 22:00:00 CDT
	2021-11-06 23:00:00 CDT
	2021-11-07 00:00:00 CDT
	2021-11-07 01:00:00 CDT
	2021-11-07 01:00:00 CST
	2021-11-07 02:00:00 CST
	2021-11-07 03:00:00 CST
	2021-11-07 04:00:00 CST
	2021-11-07 05:00:00 CST
	2021-11-07 06:00:00 CST
	2021-11-07 07:00:00 CST
	2021-11-07 08:00:00 CST
	2021-11-07 09:00:00 CST
	2021-11-07 10:00:00 CST
	2021-11-07 11:00:00 CST
	2021-11-07 12:00:00 CST
	2021-11-07 13:00:00 CST
	2021-11-07 14:00:00 CST
	2021-11-07 15:00:00 CST
	2021-11-07 16:00:00 CST
	2021-11-07 17:00:00 CST
	2021-11-07 18:00:00 CST
]
```

Failure on a `groupby`. The fail on a `groupby` seems to have been introduced in `0.16.10` as on `0.16.9` i don't get that error.
```python
t2 =pl.date_range(datetime(2021,11,7), datetime(2021,11,8), "1m", time_zone="UTC").dt.convert_time_zone("US/Central")
>>> df = pl.DataFrame({"time":t2, "value": range(1441)})
>>> df.groupby_dynamic("time", every="1h").agg(pl.col("value").mean())
thread '<unnamed>' panicked at 'Ambiguous local time, ranging from 2021-11-07T01:00:00CDT to 2021-11-07T01:00:00CST', /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/chrono-0.4.23/src/offset/mod.rs:189:17
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/***/lib/python3.10/site-packages/polars/internals/dataframe/groupby.py", line 929, in agg
    self.df.lazy()
  File "/***/lib/python3.10/site-packages/polars/internals/lazyframe/frame.py", line 1430, in collect
    return pli.wrap_df(ldf.collect())
pyo3_runtime.PanicException: Ambiguous local time, ranging from 2021-11-07T01:00:00CDT to 2021-11-07T01:00:00CST
>>> pl.__version__
'0.16.11'
```

```python
>>> t2 =pl.date_range(datetime(2021,11,7), datetime(2021,11,8), "1m", time_zone="UTC").dt.convert_time_zone("US/Central")
>>> df = pl.DataFrame({"time":t2, "value": range(1441)})
>>> df.groupby_dynamic("time", every="1h").agg(pl.col("value").mean())
shape: (25, 2)
┌──────────────────────────┬────────┐
│ time                     ┆ value  │
│ ---                      ┆ ---    │
│ datetime[μs, US/Central] ┆ f64    │
╞══════════════════════════╪════════╡
│ 2021-11-06 19:00:00 CDT  ┆ 29.5   │
│ 2021-11-06 20:00:00 CDT  ┆ 89.5   │
│ 2021-11-06 21:00:00 CDT  ┆ 149.5  │
│ 2021-11-06 22:00:00 CDT  ┆ 209.5  │
│ ...                      ┆ ...    │
│ 2021-11-07 15:00:00 CST  ┆ 1289.5 │
│ 2021-11-07 16:00:00 CST  ┆ 1349.5 │
│ 2021-11-07 17:00:00 CST  ┆ 1409.5 │
│ 2021-11-07 18:00:00 CST  ┆ 1440.0 │
└──────────────────────────┴────────┘
>>> pl.__version__
'0.16.8'


```

Sorry for the `timezone` pounding ... 

### Reproducible example

```python
from datetime import datetime
import polars as pl
t1 = pl.date_range(datetime(2021,11,7), datetime(2021,11,8), "1h", time_zone="US/Central")  # fails
t2 = pl.date_range(datetime(2021,11,7), datetime(2021,11,8), "1h", time_zone="UTC").dt.convert_time_zone("US/Central") # works as intended

t2 =pl.date_range(datetime(2021,11,7), datetime(2021,11,8), "1m", time_zone="UTC").dt.convert_time_zone("US/Central")
df = pl.DataFrame({"time":t2, "value": range(1441)})
df.groupby_dynamic("time", every="1h").agg(pl.col("value").mean())
```

### Expected behavior

I expect `t1` in the example to works as `t2`.

`2021-11-07T01:00:00CDT` and  `2021-11-07T01:00:00CST` are not ambiguous as they are different times (based on the timezone).

### Installed versions

<details>

```
Polars: 0.16.11
Index type: UInt32
Platform: macOS-13.2.1-x86_64-i386-64bit
Python: 3.10.10 (main, Feb 16 2023, 02:55:02) [Clang 14.0.0 (clang-1400.0.29.202)]
---Optional dependencies---
pyarrow: 10.0.1
pandas: 1.5.0
numpy: 1.23.3
fsspec: <not installed>
connectorx: <not installed>
xlsx2csv: <not installed>
deltalake: <not installed>
matplotlib: 3.7.0
```

</details>

### LINKED ISSUE #7496: Regression: timezone-aware date_range throws wrong results
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

`'1d'` is treated as a fixed-duration interval, which isn't necessarily the case for timezone-aware (e.g. during DST)

### Reproducible example

```python
In [3]: pl.date_range(datetime(2021,11,7), datetime(2021,11,9), "1d", time_zone="US/Central")
Out[3]:
shape: (3,)
Series: '' [datetime[μs, US/Central]]
[
        2021-11-07 00:00:00 CDT
        2021-11-07 23:00:00 CST
        2021-11-08 23:00:00 CST
]
```

### Expected behavior

```python
shape: (3,)
Series: '' [datetime[μs, US/Central]]
[
        2021-11-07 00:00:00 CDT
        2021-11-08 00:00:00 CST
        2021-11-09 00:00:00 CST
]
```

This is due to https://github.com/pola-rs/polars/pull/7430 (my bad, apologies!)

I'd suggest just reverting #7430 for now, perhaps with a clearer error message - better to error loud and clear on ambiguous dates than to deliver incorrect results. I'll put a PR together tomorrow

### Installed versions

<details>

```
---Version info---
Polars: 0.16.12
Index type: UInt32
Platform: Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Python: 3.8.16 (default, Dec  7 2022, 01:12:06)
[GCC 11.3.0]
---Optional dependencies---
pyarrow: 11.0.0
pandas: 2.0.0rc0
numpy: 1.24.2
fsspec: <not installed>
connectorx: <not installed>
xlsx2csv: <not installed>
deltalake: <not installed>
matplotlib: 3.6.2

```

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #7418
closes #7496

### CODE REVIEW

**Comment by ritchie46** [comment]
Awesome! Can you undraft. Can include it in today's release. :)

**Comment by MarcoGorelli** [comment]
sure, undrafting for now - currently it's parsing the time zone each time, did you want to take it like this and leave perf improvements to a follow-up?

**Comment by ritchie46** [comment]
Oh, right. Let's optimize directly. Can release without the patch. 👍

**Comment by MarcoGorelli** [comment]
Cool, have updated, it now parses the time zone once in `date_range_impl` and then just passes it down

For parts which don't (yet) respect the time zone properly (groupby_dynamic), I'm just passing `None` such as to not change the current behaviour. But what's introduced in this PR would make it easier to properly address groupby_dynamic and resample later

Passing `None` when a function expects `Option<&impl TimeZoneTrait>` errors, so I passed `None::<chrono::FixedOffset>` as `chrono` is typically already available in most places (I did have to add it to `polars-lazy/polars-plan` under the `date-offset` feature though)

File: polars/polars-time/src/date_range.rs [hunk 1]
```diff
@@ -1,14 +1,37 @@
...
+#[cfg(feature = "timezones")]
+use crate::windows::duration::localize_datetime;
 
 pub fn in_nanoseconds_window(ndt: &NaiveDateTime) -> bool {
     // ~584 year around 1970
     !(ndt.year() > 2554 || ndt.year() < 1386)
 }
 
+#[cfg(feature = "timezones")]
+fn localize_timestamp<T: TimeZoneTrait>(timestamp: i64, tu: TimeUnit, tz: T) -> PolarsResult<i64> {
```
Comment (ritchie46) [inline polars/polars-time/src/date_range.rs:21]: I don't think we will use this in a tight loop. If we do, we can factor out the duplicate branching and &fn indirection:

```rust
#[cfg(feature = "timezones")]
fn localize_timestamp<T: TimeZoneTrait>(timestamp: i64, tu: TimeUnit, tz: T) -> PolarsResult<i64> {
    Ok(match tu {
        TimeUnit::Nanoseconds => localize_datetime(timestamp_ns_to_datetime(timestamp), &tz)?.timestamp_nanos(),
        TimeUnit::Microseconds => localize_datetime(timestamp_us_to_datetime(timestamp), &tz)?.timestamp_micros(),
        TimeUnit::Milliseconds => localize_datetime(timestamp_ms_to_datetime(timestamp), &tz)?.timestamp_millis(),
    })
}
```

I expect that to be faster, but for now it is fine. :+1:

File: polars/polars-time/src/windows/window.rs [hunk 2]
```diff
@@ -218,9 +230,11 @@ impl BoundsIter {
                     let dt = dt.naive_utc();
                     let start = to(dt);
                     // apply the 'offset'
-                    let start = offset(&window.offset, start);
+                    // TODO remove unwrap once time zone is respected
+                    let start = offset(&window.offset, start, None::<&FixedOffset>).unwrap();
                     // and compute the end of the window defined by the 'period'
-                    let stop = offset(&window.period, start);
+                    // TODO remove unwrap once time zone is respected
+                    let stop = offset(&window.period, start, None::<&FixedOffset>).unwrap();
```
Comment (ritchie46) [inline polars/polars-time/src/windows/window.rs:237]: I think we should add a

```rust
pub const NO_TIMEZONE: Option<&FixedOffset> = None;
```

in `polars-time/src/lib.rs` to make this easier to write.

File: polars/polars-time/src/windows/window.rs [hunk 3]
```diff
@@ -247,17 +261,42 @@ impl Iterator for BoundsIter {
         if self.bi.start < self.boundary.stop {
             let out = self.bi;
             match self.tu {
+                // TODO remove unwrap once time zone is respected
```
Comment (ritchie46) [inline polars/polars-time/src/windows/window.rs:264]: Do we already have an issue for this? That means we validate timezone upon entry in the API right? I like that idea as it removes a lot of assertions downstream.

**Review by ritchie46 — COMMENTED** [review]
Thanks @MarcoGorelli. Looks good. I have left one small remark that would make it a bit easier to write.

It is not a blocker for the merge. I will include it in the (a little bit postponed) release of yesterday, today! ;)

File: polars/polars-time/src/windows/window.rs (hunk 3 above)
Comment (MarcoGorelli) [reply polars/polars-time/src/windows/window.rs:264]: There's https://github.com/pola-rs/polars/issues/7434

The time zone itself is validated upon entry (e.g. here it's validated in the entry to `date_range_impl`) - the reason for the unwrap isn't that the time zone might not be valid, but that adding a non-fixed offset might result in an ambiguous or non-existent date

For example, if you were to start on `'2021-10-07T01:00 US/Central'` and to add a month, you would end up on `'2021-11-07T01:00 US/Central'` - problem is, is that `'2021-11-07T01:00 CDT'` or `'2021-11-07T01:00 CST'`? Until https://github.com/pola-rs/polars/issues/6672 is addressed, I think the only option would be to error

File: polars/polars-time/src/date_range.rs (hunk 1 above)
Comment (MarcoGorelli) [reply polars/polars-time/src/date_range.rs:21]: it's only used once upon function entry - when were you looking to release? I have about 20 mins before starting work, I'll try to address

File: polars/polars-time/src/windows/window.rs (hunk 3 above)
Comment (ritchie46) [reply polars/polars-time/src/windows/window.rs:262]: Right, thanks for the explanation. :+1:

**Pull request merged**